proposals/lip-introduce-reward-sharing.md

```diff
@@ -0,0 +1,551 @@
+  "properties": {
+        "delegateAddress": {
+            "dataType": "bytes",
+            "length": ADDRESS_LENGTH,
+            "fieldNumber": 1
+        },
+        "oldCommission": {
+            "dataType": "uint32",
+            "fieldNumber": 2
+        },
+        "newCommission": {
+            "dataType": "uint32",
+            "fieldNumber": 3
+        }
+    }
+}
+```
+
+#### rewardsAssigned
+
+This event has `name = EVENT_NAME_REWARDS_ASSIGNED`. This event is emitted whenever rewards are assigned to a voter.
+
+##### Topics
+
+* `voterAddress` : The address of the voter receiving the rewards.
+
+##### Data
+
+```java
+rewardsAssignedEventParams = {
+  "type": "object",
+  "required": ["voterAddress", "delegateAddress", "tokenID", "amount"]
+  "properties": {
+        "voterAddress": {
+            "dataType": "bytes",
+            "length": ADDRESS_LENGTH,
+            "fieldNumber": 1
+        },
+        "delegateAddress": {
+            "dataType": "bytes",
+            "length": ADDRESS_LENGTH,
+            "fieldNumber": 2
+        },
+        "tokenID": {
+            "dataType": "bytes",
+            "length": TOKEN_ID_LENGTH,
+            "fieldNumber": 3
+        },
+        "amount": {
+            "dataType": "uint64",
+            "fieldNumber": 4
+        }
+    }
+}
+```
+
+
+### Internal Functions
+
+We define the internal functions added in the DPoS module to support reward sharing. All other internal functions of the DPoS module are defined in [LIP 0057][lip-0057#internal-functions]. 
+
+#### calculateVoteRewards
+
+This function calculates the current reward of a particular vote. The input is a `voteObject`, i.e., an object similar to the ones stored in the sentVotes array of the voter substore, containing `delegateAddress`, `amount` and `voteSharingCoefficients`.
+
+
+```python
+def calculateVoteRewards(vote: Vote,tokenID: TokenID):
+    i = index of item in delegateStore(vote.delegateAddress).sharingCoefficients with item.tokenID == tokenID
+
+    # transform variables to Q96    
+    delegateSharingCoefficient = bytesToQ96(delegateStore(vote.delegateAddress).sharingCoefficients[i].coefficient)
+    amount = Q96(vote.amount)
+    voteSharingCoefficient = bytesToQ96(vote.voteSharingCoefficients[i].coefficient)
+
+    # calculate reward
+    reward = mul_96(amount, sub_96(delegateSharingCoefficient, voteSharingCoefficient))
+
+    return Q_96_ToInt(reward)
+```
+
+Note that here `i` specifies the index of entries for token with id `tokenID` in the sharing coefficients array in both for the delegate and the voter substore. The fact that both arrays have the same `tokenID` in the same position is guaranteed from the fact that both arrays are sorted in lexicographic order. For the delegate substore array this is done in the [`updateSharedRewards`](#updatesharedrewards) function and for the voter substore in the [assignVoteRewards](#assignvoterewards) function.
+
+#### assignVoteRewards
```

Comment:
Should we align the pseudocode with the implementation?

Reply:
We decided finally not to do the update in the LIP (I made the change and then revert it back to the previous state) since the specifications look much cleaner if it is clear that the state change should happen at every call of this function.  I think it is fine if for implementation reasons you slightly modify this, as long as the logic is the same.